### PR TITLE
Implement narrow from a regular tensor to jagged tensor

### PR DESCRIPTION
**Pull request by ani300**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #112770

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: torch/nested/__init__.py [hunk 1]
```diff
@@ -186,3 +186,34 @@ def nested_tensor(tensor_list, *, dtype=None, layout=None, device=None, requires
         return nt
     else:
         raise RuntimeError(f"Specified layout is unsupported for nested tensors: {layout}")
+
+
+
+
+def narrow(tensor: Tensor, dim: int, start: Union[SymInt, Tensor], length: Union[SymInt, Tensor], layout=torch.jagged):
```
Comment (cpuhrsch) [inline torch/nested/__init__.py:193]: I don't think we should add a layout argument here. It doesn't follow the narrow syntax for regular tensors. If the user wants to change the layout they can use `to`. Since this is a view operation it shouldn't affect performance. Furthermore, layout change is likely not support when this is view only.

Comment (jbschlosser) [reply torch/nested/__init__.py:193]: > I don't think we should add a layout argument here. It doesn't follow the narrow syntax for regular tensors.

This is true, but we're also not limited to the original `narrow()` syntax since this is a new function in the `torch.nested` namespace.

So I sort of disagree. It's possible to create a view over a dense buffer with both the expanded jagged layout and the strided layout, and I think the `layout` arg is a natural extension for indicating which type of NT + metadata is to be constructed.

> If the user wants to change the layout they can use to.

Current semantics are such that this will copy, no? I don't quite see how we can guarantee `torch.nested.narrow(...).to(layout=...)` is non-copying if it's the difference between a C++ NT and a python subclass jagged layout NT.

Comment (cpuhrsch) [reply torch/nested/__init__.py:193]: `torch.nested.narrow(...).to(layout=...)` can be made non-copying via torch.compile and pattern matching. 

But you're right that it's a new function. Maybe in the future we can extend torch.narrow with similar functionality.

Alright, I'm convinced.

**Review by cpuhrsch — APPROVED** [review]

File: torch/nested/_internal/nested_tensor.py [hunk 2]
```diff
@@ -285,5 +304,36 @@ def jagged_from_list(
...
+    """Constructs a NestedTensor backed by jagged layout from a tensor, starts of sequences, and sequence lengths"""
+    batch_size = tensor.shape[0]
+    try:
+        start_list = starts.expand(batch_size, 1)
+        length_list = lengths.expand(batch_size, 1)
+    except RuntimeError as e:
+        raise RuntimeError(
+            "When constructing a jagged nested tensor using narrow(), "
+            "your start and length must be a Tensor that broadcasts to input.shape[0] x 1"
+        ) from e
```
Comment (soulitzer) [inline torch/nested/_internal/nested_tensor.py:319]: You can use `is_expandable_to` in `torch/_prims_common/__init__.py` instead

File: torch/nested/_internal/nested_tensor.py [hunk 3]
```diff
@@ -232,6 +240,17 @@ def backward(ctx, gO: NestedTensor):  # type: ignore[override]
         return gO.values(), None, None
 
 
+# Not actually a view!
+class ViewNoncontiguousNestedFromBuffer(torch.autograd.Function):
```
Comment (soulitzer) [inline torch/nested/_internal/nested_tensor.py:244]: nit:
```suggestion
class ViewNonContiguousNestedFromBuffer(torch.autograd.Function):
```
is slightly more readable to me at a glance

File: torch/nested/_internal/nested_tensor.py [hunk 4]
```diff
@@ -285,5 +304,36 @@ def jagged_from_list(
...
+        start_list = starts.expand(batch_size, 1)
+        length_list = lengths.expand(batch_size, 1)
+    except RuntimeError as e:
+        raise RuntimeError(
+            "When constructing a jagged nested tensor using narrow(), "
+            "your start and length must be a Tensor that broadcasts to input.shape[0] x 1"
+        ) from e
+
+    # Calculate jagged offsets
+    max_seq_len = tensor.shape[1]
```
Comment (soulitzer) [inline torch/nested/_internal/nested_tensor.py:322]: Add explicit assertion that dim >= 2

File: torch/nested/_internal/ops.py [hunk 5]
```diff
@@ -36,7 +36,8 @@ def check_schema(schema_str: str, func, *args, **kwargs) -> None:
 
     arg_type_check_fns = {
         "t": lambda x: isinstance(x, torch.Tensor) and not isinstance(x, NestedTensor),
-        "jt": lambda x: isinstance(x, NestedTensor),
+        "jt": lambda x: isinstance(x, NestedTensor) and x._lengths is None,
+        "jt_nc": lambda x: isinstance(x, NestedTensor),
```
Comment (soulitzer) [inline torch/nested/_internal/ops.py:40]: This new type won't be too useful today as you won't actually be able to register two funcs to the same aten op overload even if those ops have different schemas. 

For now, what you probably want to do is just to branch inside whatever op you are trying to implement.

Perhaps in the future we want to go the route of writing a general dispatching mechanism, I'm not sure. Or its also entirely possible that we'd revert the t vs jt distinction as well. We don't have many ops implemented, so I think its too early to decide today.

File: torch/nested/__init__.py [hunk 6]
```diff
@@ -186,3 +186,34 @@ def nested_tensor(tensor_list, *, dtype=None, layout=None, device=None, requires
...
+    if layout == torch.jagged:
+        from torch.nested._internal.nested_tensor import jagged_from_tensor_and_lengths
+
+        if isinstance(start, SymInt):
+            start = torch.tensor([int(start)])
+
+        if isinstance(length, SymInt):
+            length = torch.tensor([int(length)])
+
+        nt = jagged_from_tensor_and_lengths(tensor, start, length)
```
Comment (soulitzer) [inline torch/nested/__init__.py:214]: nit: You actually returned 3 things from this function, but the name suggests that you only expected to return one thing.

File: torch/nested/_internal/nested_tensor.py [hunk 7]
```diff
@@ -21,6 +21,7 @@ def get_tensor_id(tensor, *, coeff=1):
 class NestedTensor(torch.Tensor):
     _values: torch.Tensor  # type: ignore[assignment]
     _offsets: torch.Tensor
+    _lengths: Optional[torch.Tensor]
```
Comment (soulitzer) [inline torch/nested/_internal/nested_tensor.py:24]: What do we expect the .sizes() to return for this tensor?

File: torch/nested/__init__.py (hunk 6 above)
Comment (soulitzer) [reply torch/nested/__init__.py:214]: Do we actually want to return all 3 to the user?

File: torch/nested/_internal/ops.py (hunk 5 above)
Comment (ani300) [reply torch/nested/_internal/ops.py:40]: ah... I was trying to get out of adding an if statement to every single function we implement, but instead I will change the register_func call to accept a parameter of whether non-contiguous tensors are allowed or not (given most functions won't accept them at all)

File: torch/nested/_internal/nested_tensor.py (hunk 7 above)
Comment (ani300) [reply torch/nested/_internal/nested_tensor.py:24]: this should be the same size as `_offsets`

File: torch/nested/__init__.py (hunk 6 above)
Comment (ani300) [reply torch/nested/__init__.py:214]: yeah, I'll change it, I was trying to keep it similar to `jagged_from_list` which returns its offsets and is semantically a very similar function

File: torch/nested/_internal/nested_tensor.py (hunk 7 above)
Comment (soulitzer) [reply torch/nested/_internal/nested_tensor.py:24]: Oh what I mean is what the .size() is for the NestedTensor
Currently, given a tensor with real size of [3, [1, 2, 3], 2]], values is represented as [1+2+3, 2] = [6, 2].
Today, in order to represent the ragged [1, 2, 3] dimension, we have a special int which we call "singleton int" returned in the size, e.g. [3, j0, 2] where the singleton int  j0 corresponds 1:1 with offsets Tensor objects. 
The semantics of this special int are basically that the main supported operation is `eq` and most other operations like `add` are disallowed. The idea is to allow this int to represent a kind of id for the offset and allow size comparison in O(1). 
Introducing lengths means that we might want a second singleton int corresponding 1:1 with lengths, and that we want to return that singleton int instead. 
There's some more context in `c10/core/SingletonSymNodeImpl.h`

File: torch/nested/__init__.py (hunk 6 above)
Comment (soulitzer) [reply torch/nested/__init__.py:214]: Yeah I think its okay for the helper function `jagged_from_tensor_and_lengths` to also return 3, but we don't necessarily want to return all 3 in the public API.

File: torch/nested/_internal/nested_tensor.py (hunk 7 above)
Comment (ani300) [reply torch/nested/_internal/nested_tensor.py:25]: I see... I'll change it then

File: torch/nested/__init__.py [hunk 8]
```diff
@@ -186,3 +186,59 @@ def nested_tensor(tensor_list, *, dtype=None, layout=None, device=None, requires
...
+    tensor (:class:`torch.Tensor`): a strided tensor, which will be used as the underlying data
+    for the nested tensor if using the jagged layout or will be copied for the strided layout.
+    dim (int): the dimension where narrow will be applied. Only `dim=1` is supported for the
+    jagged layout, while strided supports all dim
+    start (Union[int, :class:`torch.Tensor`]): starting element for the narrow operation
+    length (Union[int, :class:`torch.Tensor`]): number of elements taken during the narrow op
+
+Keyword arguments:
+    layout (:class:`torch.layout`, optional): the desired layout of returned nested tensor.
+        Only strided and jagged layouts are supported. Default: if None, the jagged layout.
```
Comment (jbschlosser) [inline torch/nested/__init__.py:210]: nit: None default should indicate strided, as is consistent with the behavior in other places (e.g. `nested_tensor()` / `as_nested_tensor()`).

also is None being handled? I didn't see it but I may have just missed it

**Review by jbschlosser — COMMENTED** [review]
Nice work! I'm curious to see how this looks with SDPA :)

Do you mind adding a few tests into `test/test_nestedtensor.py`? I think we can check that things nicely error out for all ops when they encounter a non-contiguous jagged NT.

Along that line: I think we want to expand the definition of `contiguous()` to take `lengths` into account + add a corresponding test.

File: torch/nested/_internal/nested_tensor.py (hunk 7 above)
Comment (jbschlosser) [reply torch/nested/_internal/nested_tensor.py:25]: @soulitzer good points! The increase in complexity scares me though.

Does it make sense to expand the current singleton symint to represent the combination of `(offsets, lengths)` when both are present?

File: torch/nested/_internal/nested_tensor.py [hunk 9]
```diff
@@ -285,5 +305,41 @@ def jagged_from_list(
...
+    max_seq_len = tensor.shape[1]
+    offset_lengths = max_seq_len * torch.arange(
+        0, batch_size, dtype=torch.int64, device=tensor.device
+    )
+    # Jagged layout specifies that offsets are stored as int64 on the same device as values.
+    offsets = start_list + offset_lengths
+
+    # Reshape buffer to flatten the 1st and 2nd dimension
+    if len(tensor.shape) > 2:
+        values = tensor.reshape(-1, *tensor.shape[2:])
```
[the remark this answers is not in the document]
Comment (soulitzer) [reply torch/nested/_internal/nested_tensor.py:337]: hmm this makes me think whether we should make assertions about the contiguity of the input tensor. 
both reshape and flatten can conditionally copy depending whether the input shape is cleanly reshapable to the desired shape. 
I feel like we at least do something like view to make users explicitly contiguify if necessary to avoid any silent performance cliffs when inputs change.

[the remark this answers is not in the document]
Comment (jbschlosser) [reply torch/nested/_internal/nested_tensor.py:337]: > I feel like we at least do something like view to make users explicitly contiguify if necessary to avoid any silent performance cliffs when inputs change.

probably a good idea to require being explicit about this, I agree

File: torch/nested/_internal/nested_tensor.py (hunk 7 above)
Comment (jbschlosser) [reply torch/nested/_internal/nested_tensor.py:25]: from offline discussion: when `lengths` exist, we should use those to produce the singleton symint instead of using `offsets`. Two jagged layout NTs with the same `lengths` will have the same singleton symint for that shape, and theoretically are compatible for binary op invocation. The kernel that performs the binary op must take into account both `offsets` and `lengths` to do the computation properly.

File: torch/nested/__init__.py (hunk 8 above)
Comment (ani300) [reply torch/nested/__init__.py:210]: it is being handled with a RuntimeError

File: torch/nested/__init__.py [hunk 10]
```diff
@@ -186,3 +186,59 @@ def nested_tensor(tensor_list, *, dtype=None, layout=None, device=None, requires
...
         raise RuntimeError(f"Specified layout is unsupported for nested tensors: {layout}")
+
+
+def narrow(tensor: Tensor, dim: int, start: Union[int, Tensor], length: Union[int, Tensor], layout=torch.jagged):
+    r"""
+Constructs a nested tensor (which might be a view) from :attr:`tensor`, a strided tensor. This follows
+similar semantics ot torch.Tensor.narrow, where in the :attr:`dim`-th dimension the new nested tensor
+(maybe view) shows only the elements in the interval `[start, start+length]`. As nested representations
+allow for a different `start` and `length` at each 'row' of that dimension, :attr:`start` and :attr:`length`
+can also be tensors of shape `tensor.shape[0] x 1`.
```
[the remark this answers is not in the document]
Comment (ani300) [reply torch/nested/__init__.py:197]: changing to tensor.shape[0], makes more sense

File: torch/nested/__init__.py [hunk 11]
```diff
@@ -186,3 +186,59 @@ def nested_tensor(tensor_list, *, dtype=None, layout=None, device=None, requires
...
+
+Example::
+
+    >>> a = torch.arange(3, dtype=torch.float, requires_grad=True)
+    >>> b = torch.arange(5, dtype=torch.float, requires_grad=True)
+    >>> nt = torch.nested.nested_tensor([a, b], requires_grad=True)
+    >>> nt.is_leaf
+    True
+    """
+    if not isinstance(start, int) or isinstance(start, Tensor):
```
[the remark this answers is not in the document]
Comment (ani300) [reply torch/nested/__init__.py:220]: added SymInt check

File: torch/nested/_internal/ops.py (hunk 5 above)
Comment (ani300) [reply torch/nested/_internal/ops.py:40]: on second thought, "jt_nc" might not be the best name, but it is the one that allows both the contiguous and noncont. versions to pass though to the actual kernel code vs "jt" only letting the contiguous ones pass to maintain compatibility

Comment (jbschlosser) [reply torch/nested/_internal/ops.py]: yeah I'm okay with your initial approach. I did something similar to that locally when I was playing around with this and it minimized the changes needed to existing op impls

File: torch/nested/_internal/ops.py [hunk 12]
```diff
@@ -36,7 +36,8 @@ def check_schema(schema_str: str, func, *args, **kwargs) -> None:
 
     arg_type_check_fns = {
         "t": lambda x: isinstance(x, torch.Tensor) and not isinstance(x, NestedTensor),
-        "jt": lambda x: isinstance(x, NestedTensor),
+        "jt": lambda x: isinstance(x, NestedTensor) and x._lengths is None,
```
Comment (jbschlosser) [inline torch/nested/_internal/ops.py:39]: nit: maybe add a small comment describing that `jt` requires contiguous JTs only

File: torch/nested/__init__.py [hunk 13]
```diff
@@ -186,3 +187,63 @@ def nested_tensor(tensor_list, *, dtype=None, layout=None, device=None, requires
         return nt
     else:
         raise RuntimeError(f"Specified layout is unsupported for nested tensors: {layout}")
+
+
+def narrow(tensor: Tensor, dim: int, start: Union[int, Tensor], length: Union[int, Tensor], layout=torch.strided) -> Tensor:
+    r"""
+Constructs a nested tensor (which might be a view) from :attr:`tensor`, a strided tensor. This follows
+similar semantics to torch.Tensor.narrow, where in the :attr:`dim`-th dimension the new nested tensor
+(maybe view) shows only the elements in the interval `[start, start+length]`. As nested representations
```
Comment (jbschlosser) [inline torch/nested/__init__.py:196]: wording / formatting nit: rather than mentioning views in-line here, I think it'd be clearer to talk about view / copy behavior for each layout in a separate paragraph after this one

**Review by jbschlosser — COMMENTED** [review]
Looking really nice, thanks for all your hard work!

I'm just concerned about one thing: the format of `offsets`. Traditionally, the jagged layout defines it to be of shape `B + 1` for batch size `B`, and it includes the final offset (e.g. `offsets=[0, 3, 5, 7]` with B=3 and a values buffer of shape `[7, D]`). For the contiguous case, this allows for lengths to be computed with `offsets.diff()`.

From what I can tell, this PR redefines `offsets` to leave off the final offset, resulting in `offsets.shape[0] == lengths.shape[0] == B`.

I don't think we can change this for the contiguous case, as it's well-established at this point. For consistency, I'd also argue that the offsets shape should be the same across non-contiguous and contiguous cases. I admit this is a little strange, as the final offset doesn't have a ton of utility when `lengths` are defined separately.

@soulitzer / @cpuhrsch / @drisspg: what are your opinions on this?

**Comment by jbschlosser** [comment]
from offline discussion: we're in agreement that `offsets` should consistently be in the old format: shape `B + 1`

**Comment by ani300** [comment]
I've updated all the offsets, hopefully I didn't miss any, but I can't seem to figure out the issue with the recompile happening on dynamo that breaks CI

File: torch/nested/__init__.py [hunk 14]
```diff
@@ -187,3 +188,69 @@ def nested_tensor(tensor_list, *, dtype=None, layout=None, device=None, requires
         return nt
     else:
         raise RuntimeError(f"Specified layout is unsupported for nested tensors: {layout}")
+
+
+def narrow(tensor: Tensor, dim: int, start: Union[int, Tensor], length: Union[int, Tensor], layout=torch.strided) -> Tensor:
```
Comment (cpuhrsch) [inline torch/nested/__init__.py:193]: Hm, did the documentation for this make it onto the website?

Comment (ani300) [reply torch/nested/__init__.py:193]: I might have forgotten to commit the changes to the docs adding this function, as I don't see it in the commit
